### PR TITLE
Split core/str/mod.rs to smaller files

### PR DESCRIPTION
**Pull request by tesuji**
Note for reviewer:
* I split to multiple commits for easier reviewing, but I could git squash them all to one if requested.
* Recommend pulling this change locally and using advanced git diff viewer or this command:
  ```bash
  git show --reverse --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space master..
  ```

---

I split `core/str/mod.rs` to these modules:

* `converts`: Contains helper functions to convert from bytes to str.
* `error`: For error structs like Utf8Error.
* `iter`: For iterators of many str methods.
* `traits`: For indexing operations and build in traits on str.
* `validations`: For functions validating utf8 --- This name is awkward, maybe utf8.rs is better.

### CODE REVIEW

**Comment by lzutao** [comment]
Rebased.

**Comment by lzutao** [comment]
Rebased.

**Comment by varkor** [comment]
@rust-lang/libs: is there someone who could review this PR? It bitrots quite quickly.

@bors p=100

**Comment by Amanieu** [comment]
@bors r+

Also I learned something new about git today!

**Comment by bjorn3** [comment]
This caused a 26.6% reduction on clap-rs-debug incr-patched: println. https://perf.rust-lang.org/compare.html?start=ef663a8a48ea6b98b43cbfaefd99316b36b16825&end=9bb55dc8642d811d66a7599812009cc063577e00&stat=instructions:u

**Comment by lzutao** [comment]
It also regresses wall-time and some max-rss tests: https://perf.rust-lang.org/compare.html?start=ef663a8a48ea6b98b43cbfaefd99316b36b16825&end=9bb55dc8642d811d66a7599812009cc063577e00&stat=wall-time

I don't think we could do anything to improve it.

**Comment by SimonSapin** [comment]
@bjorn3 Is that number really due to this PR? It looks like that benchmark has been oscillating between two results, recently:

image

**Comment by bjorn3** [comment]
Ah, ok.

File: library/core/src/str/traits.rs [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+/// }
+///
+/// let p = Point::from_str("(1,2)");
+/// assert_eq!(p.unwrap(), Point{ x: 1, y: 2} )
+/// ```
+#[stable(feature = "rust1", since = "1.0.0")]
+pub trait FromStr: Sized {
+    /// The associated error which can be returned from parsing.
+    #[stable(feature = "rust1", since = "1.0.0")]
+    type Err;
```
Comment (nervenes) [inline library/core/src/str/traits.rs:537]: i think this was supposed to be ``type Error``, is it possible to change this?

Comment (sfackler) [reply library/core/src/str/traits.rs:537]: You are 9.5 years too late to change this.

Comment (bjorn3) [reply library/core/src/str/traits.rs:537]: To elaborate: It was already `type Err` before this PR. https://github.com/rust-lang/rust/pull/21718 which stabilized `FromStr` introduced `type Err`.

**Pull request merged**